x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -136,53 +146,221 @@ protected void readInternal(ByteBuffer b) throws IOException {
-                    final Tuple<Long, Long> rangeToRead = Tuple.tuple(pos, Math.min(pos + len, rangeToWrite.v2()));
-
-                    bytesRead = cacheFile.fetchAsync(rangeToWrite, rangeToRead, (channel) -> {
-                        final int read;
-                        if ((rangeToRead.v2() - rangeToRead.v1()) < b.remaining()) {
-                            final ByteBuffer duplicate = b.duplicate();
-                            duplicate.limit(duplicate.position() + Math.toIntExact(rangeToRead.v2() - rangeToRead.v1()));
-                            read = readCacheFile(channel, pos, duplicate);
-                            assert duplicate.position() <= b.limit();
-                            b.position(duplicate.position());
-                        } else {
-                            read = readCacheFile(channel, pos, b);
-                        }
-                        return read;
-                    }, this::writeCacheFile, directory.cacheFetchAsyncExecutor()).get();
+        logger.trace("readInternal: read [{}-{}] from [{}]", position, position + length, this);
+
+        // We can detect that we're going to read the last 16 bytes (that contains the footer checksum) of the file. Such reads are often
+        // executed when opening a Directory and since we have the checksum in the snapshot metadata we can use it to fill the ByteBuffer.
+        if (length == CodecUtil.footerLength() && isClone == false && position == fileInfo.length() - length) {
+            if (readChecksumFromFileInfo(b)) {
+                logger.trace("read footer of file [{}] at position [{}], bypassing all caches", fileInfo.physicalName(), position);
+                return;
+            }
+            assert b.remaining() == length;
+        }
+
+        final List<Tuple<Long, Long>> indexCacheMisses;
+
+        // We prefer to use the index cache if the recovery is not done yet
+        if (directory.isRecoveryDone() == false) {
+            // We try to use the snapshot blob cache if:
+            // - we're reading the first N bytes of the file
+            final boolean isStartOfFile = (position + length <= BlobStoreCacheService.DEFAULT_SIZE);
+            // - the file is small enough to be fully cached in the blob cache
+            final boolean canBeFullyCached = fileInfo.length() <= BlobStoreCacheService.DEFAULT_SIZE * 2;
+
+            if (canBeFullyCached || isStartOfFile) {
+                final CachedBlob cachedBlob = directory.getCachedBlob(fileInfo.physicalName(), 0L, length);
+                if (cachedBlob != null) {
+                    logger.trace(
+                        "reading [{}] bytes of file [{}] at position [{}] using index cache",
+                        length,
+                        fileInfo.physicalName(),
+                        position
+                    );
+                    b.put(BytesReference.toBytes(cachedBlob.bytes().slice(Math.toIntExact(position), length)));
```

Comment:
Do we want to put the file during recovery into the node-local cache (at least if it's the full file)? Can be done in follow-up ofc.

Reply:
Mmm maybe, but it does introduce a good deal of troublesome interplay with cache eviction and other reads. We would have to acquire the cache file, possibly evicting other files, in order to put the data on disk. We may never need it again after the shard started up too.

Reply:
Ok yes we need to do this, or else we start the shard without all this stuff in node-local cache and then have to fetch them on the first search since we ignore the index cache once recovery is finished.